### ISSUE TITLE
misc/cgo/testcshared failure, linking to wrong libgo.so

### ISSUE THREAD

**Issue by hartzell** [golang/go#17986]
**TL;DR**: I needed to change `libgo.so` to `./libgo.so` on one line in `test.bash` to make `misc/cgo/testcshared` pass and am looking for feedback.

Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?

Building 1.7.3 with a go built from release-branch.go1.4

### What operating system and processor architecture are you using (`go env`)?

`go env` output from the successful build mentioned below:

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH=""
GORACE=""
GOROOT="/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/go-1.7.3-wzhabet756id3kyoqr6v6hnrwopfkpic"
GOTOOLDIR="/rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-5.4.0/go-1.7.3-wzhabet756id3kyoqr6v6hnrwopfkpic/pkg/tool/linux_amd64"
CC="/rss/spack-20161101/lib/spack/env/gcc/gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0"
CXX="/rss/spack-20161101/lib/spack/env/gcc/g++"
CGO_ENABLED="1"
```

I'm working in a DigitalOcean droplet based on their CentOS7 image with a minimal set of packages added.  I've worked through the problem described in #16283 by adding the kernel boot flag.  I also had to add `glibc-static` to my "minimal" set of packages so that the cgo static linking tests pass.

### What did you do?

I'm updating go packaging in the [spack](https://github.com/LLNL/spack) package manager.

I've updated their go-bootstrap package to use `release-branch.go1.4` instead of 1.4.2 to pick up Brad's [timezone fix](https://github.com/golang/go/commit/69ba71e4d5e2eeb7c55d418c86b528305edbc0fa).  I have not yet submitted a pull request to the spack project for this change (but will).

I'm using that go-bootstrap package to build 1.7.3.

I've updated their [go package](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/go/package.py)  to include an entry for the 1.7.3 tag.

The testing step has been failing in `misc/cgo/testcshared` with this message:

> ./testp: symbol lookup error: ./testp: undefined symbol: DidInitRun

After drilling down into the tests to the point where I was able to build the test itself and replicate the failure, I noticed that `testp` was linking against the "wrong" `libgo.so`.  When I added a `"./"` to the shared library name, it linked against the "right" `libgo.so` the test passed.

```
[hartzelg@spack-0 testcshared]$ $(go env CC) ${GOGCCFLAGS} -I ${installdir} -o testp main0.c libgo.$libext
[hartzelg@spack-0 testcshared]$ ./testp
./testp: symbol lookup error: ./testp: undefined symbol: DidInitRun
[hartzelg@spack-0 testcshared]$ ldd testp
	linux-vdso.so.1 =>  (0x00007ffcd83c1000)
	libgo.so => /rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-5ugmutg4tsvybv3zsrhfa5h67ltzrc5i/lib64/libgo.so (0x00007fdffe6aa000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fdffe487000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fdffe0c5000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fdffddc3000)
	libgcc_s.so.1 => /rss/spack-20161101/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gcc-5.4.0-5ugmutg4tsvybv3zsrhfa5h67ltzrc5i/lib64/libgcc_s.so.1 (0x00007fdffdbab000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fdfff8ea000)
[hartzelg@spack-0 testcshared]$ pwd
/rss/spack-20161101/var/spack/stage/go-1.7.3-wzhabet756id3kyoqr6v6hnrwopfkpic/go/misc/cgo/testcshared
[hartzelg@spack-0 testcshared]$ $(go env CC) ${GOGCCFLAGS} -I ${installdir} -o testp main0.c ./libgo.$libext
[hartzelg@spack-0 testcshared]$ ldd testp
	linux-vdso.so.1 =>  (0x00007fffa87d8000)
	./libgo.so (0x00007f520d4d3000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f520d2b0000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f520ceee000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f520d7da000)
[hartzelg@spack-0 testcshared]$ ./testp
PASS
```

If I patch the `test.bash` script like so:

```
--- misc/cgo/testcshared/test.bash.orig	2016-11-19 00:00:11.917000000 +0000
+++ misc/cgo/testcshared/test.bash	2016-11-19 00:00:22.081000000 +0000
@@ -107,7 +107,7 @@

 # test0: exported symbols in shared lib are accessible.
 # TODO(iant): using _shared here shouldn't really be necessary.
-$(go env CC) ${GOGCCFLAGS} -I ${installdir} -o testp main0.c libgo.$libext
+$(go env CC) ${GOGCCFLAGS} -I ${installdir} -o testp main0.c ./libgo.$libext
 binpush testp

 output=$(run LD_LIBRARY_PATH=. ./testp)
```

then I can build 1.7.3 successfully.

Since the 1.7.3 release has been in the wild for a while and I don't see any other issues about this I suspect it's related to the gymanstics that Spack does to support multiple installations of compilers and libraries and etc....

I'm inclined to add a patch to `test.bash` to Spack's install step (as above) but I'd like to get feedback from more experienced gophers first.

**Comments?**

**Comment by minux** [comment]
It seems to be caused by naming conflict with libgo.so provided by gccgo.
However, why is libgo.so from gccgo lacks a version in the file name?

**Comment by ianlancetaylor** [comment]
It's normal to install a file named libgo.so, since the linker will look for an extension of ".so" when resolving a -lgo option.  It will be a symlink to the real file with the version number.

I'll add the `./`, it's the right thing to do anyhow.
